OSDOCS-6248:adds AdminNetworkPolicy to OVN-K (#70800)
openshift/openshift-docs: networking/ovn_kubernetes_network_provider/ovn-k-network-policy.adoc

:_mod-docs-content-type: ASSEMBLY
[id="ovn-k-network-policy"]
= OVN-Kubernetes network policy
include::_attributes/common-attributes.adoc[]
:context: ovn-k-network-policy

toc::[]

include::snippets/technology-preview.adoc[]

> OVN-Kubernetes network policy uses a 3 tiered Access Control List (ACLs) to evaluate and apply network policy. Network policies are evaluate in descending order from tier 1 to tier 3.

Reviewer: So `NetworkPolicy` and `AdminNetworkPolicy` are two different CRDs, features and so giving them the wider umbrella of them all being "NetworkPolicy" makes it confusing.

I would change this to something along the lines of, Kubernetes offers two features that users can use to enforce network security. One of them is the well-known NetworkPolicy API which has been around for a long time. This was designed for application developers and namespace tenants and detailed docs on that can be found <link-to-ovnk-netpol-docs> and the new feature called `AdminNetworkPolicy` allows cluster administrators to enforce non-overridable policies that take precedence over NetworkPolicies.

Something like that is better,,,? Then you can talk how "How OVNKubernetes implements these Kubernetes native features using tiered ACLs -> link to OVN's ACL docs so that those who don't know what that is can read upon that. This feels odd to drop the ball on acl, tier etc without setting some stage on the introduction paragraph...? Start with Kubernetes, spend a few seconds on what these features do generally and then go into OVNK details of ACLs and tiers

Tier 1 evaluates `AdminNetworkPolicies` (ANP). Tier 2 evaluates `NetworkPolicy`. Tier 3 evaluates `BaselineNetworkPolicy` (BANP).

> //possibly putting an image here to help describe ACLs

Reviewer: +1 yes image helps

> 

Reviewer: We should also introduce BANP here.... like two resources part of ANP feature

:_mod-docs-content-type: ASSEMBLY
[id="ovn-k-network-policy"]
= OVN-Kubernetes network policy
include::_attributes/common-attributes.adoc[]
:context: ovn-k-network-policy

toc::[]

:FeatureName: The `AdminNetworkPolicy` resource
include::snippets/technology-preview.adoc[]

Kubernetes offers two features that users can use to enforce network security. One feature that allows users to enforce network policy is the `NetworkPolicy` API that is designed mainly for application developers and namespace tenants to protect their namespaces by creating namespace-scoped policies. For more information, see xref:../../networking/network_policy/about-network-policy.adoc#nw-networkpolicy-about_about-network-policy[About network policy].

> The second feature is `AdminNetworkPolicy` which is comprised of two API: the `AdminNetworkPolicy` (ANP) API and the `BaselineAdminNetworkPolicy` (BANP) API. Together ANP and BANP are designed for cluster and network administrators to protect their entire cluster by creating cluster-scoped policies. Cluster administrators can use ANPs and BANPs to enforce non-overridable policies that take precedence over `NetworkPolicy` objects.

Author: @tssurya and @asood-rh I think we need a more generic name when we talk about "Admin network policy" in the general sense; for example, here as a feature. What do you think of "global network policy"?

Reviewer: no we can't put a new name at this stage, I think its important to prefix/suffix the terms feature and API whenever needed but mostly it should be implicitly understood...
perhaps @astoycos has better ideas here since this is important from marketing perspective of our community and project moving forward and we don't want to be stuck with a branding we don't like,

> 

Reviewer 2: Cluster administrators can use ANPs  to enforce non-overridable policies that take precedence over `NetworkPolicy` objects. This statement is true.
BANP is overridable.

OVN-Kubernetes CNI in {product-title} implements these network policies using Access Control List (ACLs) Tiers to evaluate and apply them. ACLs are evaluated in descending order from Tier 1 to Tier 3.

> Tier 1 evaluates `AdminNetworkPolicies` (ANP). Tier 2 evaluates `NetworkPolicy` policy. Tier 3 evaluates `BaselineAdminNetworkPolicy` (BANP).

Reviewer: ```suggestion
Tier 1 evaluates `AdminNetworkPolicy` (ANP). Tier 2 evaluates `NetworkPolicy` policy. Tier 3 evaluates `BaselineAdminNetworkPolicy` (BANP).
```
?
consistency :) either go plural all the way or stick to singular all the way

:_mod-docs-content-type: ASSEMBLY
[id="ovn-k-network-policy"]
= OVN-Kubernetes network policy
include::_attributes/common-attributes.adoc[]
:context: ovn-k-network-policy

toc::[]

> :FeatureName: `AdminNetworkPolicy`

Reviewer 3: Maybe:
```suggestion
:FeatureName: The `AdminNetworkPolicy` resource
```

include::snippets/technology-preview.adoc[]

> Kubernetes offers two features that users can use to enforce network security. One feature that allows users to enforce network policy is the `NetworkPolicy` API that is designed for application developers and namespace tenants. For more information, see xref:../../networking/network_policy/about-network-policy.adoc#nw-networkpolicy-about_about-network-policy[About network policy].

Author: I like the simpler prose that don't sound like jargon and these are features so I've updated with present tense.

> The second feature is `AdminNetworkPolicy` (ANP) and `BaselineAdminNetworkPolicy` (BANP) APIs. Cluster administrators can use ANP and BANP to enforce non-overridable policies that take precedence over `NetworkPolicies`.

Reviewer 3: Maybe?
```suggestion
The second feature is `AdminNetworkPolicy` (ANP) and `BaselineAdminNetworkPolicy` (BANP) APIs. Cluster administrators can use ANPs and BANPs to enforce non-overridable policies that take precedence over `NetworkPolicies`.
```

Author: This is tricky situation with plurals. You can only have one BANP so I don't think that can ever be plural. I'm going leave singular for clarity.

> The second feature is `AdminNetworkPolicy` (ANP) and `BaselineAdminNetworkPolicy` (BANP) APIs. Cluster administrators can use ANP and BANP to enforce non-overridable policies that take precedence over `NetworkPolicies`.

Author: I _think_ this is good for the RN but I don't want to discuss the newness or timeline of these in the technical documents.

> The second feature is `AdminNetworkPolicy` (ANP) and `BaselineAdminNetworkPolicy` (BANP) APIs. Cluster administrators can use ANP and BANP to enforce non-overridable policies that take precedence over `NetworkPolicies`.

Reviewer: sure feel free to cut the new (edited the comment again)

> OVN-Kubernetes implements these network policies usings a 3 tiered Access Control List (ACLs) to evaluate and apply them. Network policies are evaluate in descending order from tier 1 to tier 3.

Reviewer 3: ```suggestion
OVN-Kubernetes implements these network policies using a 3 tiered Access Control List (ACLs) to evaluate and apply them. Network policies are evaluated in descending order from tier 1 to tier 3.
```

> OVN-Kubernetes implements these network policies usings a 3 tiered Access Control List (ACLs) to evaluate and apply them. Network policies are evaluate in descending order from tier 1 to tier 3.

Author: I can't seem to find my way back to OVN man page for ACL. Do you have the link?

> OVN-Kubernetes implements these network policies usings a 3 tiered Access Control List (ACLs) to evaluate and apply them. Network policies are evaluate in descending order from tier 1 to tier 3.

Reviewer: what do you mean? It's the first thing that shows up on google: https://man7.org/linux/man-pages/man8/ovn-nbctl.8.html#COMMANDS

Tier 1 evaluates `AdminNetworkPolicies` (ANP). Tier 2 evaluates `NetworkPolicy`. Tier 3 evaluates `BaselineNetworkPolicy` (BANP).

//possibly putting an image here to help describe ACLs

//With an image this will provide a more detailed picture of the integration between ANP and BANP with NP API.

> If traffic matches an ANP rule, the rules in that ANP will be evaluated first. If the match is an ANP `allow` or `deny` rule, any existing `NetworkPolicies` and `BaselineAdminNetworkPolicy` (BANP) in the cluster will be intentionally skipped from evaluation. If the match is an ANP `pass` rule, then evaluation moves from tier 1 of the ACLs to tier 2 where `NetworkPolicy` is evaluated.

Reviewer 3: ```suggestion
If traffic matches an ANP rule, the rules in that ANP will be evaluated first. If the match is an ANP `allow` or `deny` rule, any existing `NetworkPolicies` and `BaselineAdminNetworkPolicy` (BANP) in the cluster will be intentionally skipped from evaluation. If the match is an ANP `pass` rule, then evaluation moves from tier 1 of the ACLs to tier 2 where the `NetworkPolicy` policy is evaluated.
```

> If traffic matches an ANP rule, the rules in that ANP will be evaluated first. If the match is an ANP `allow` or `deny` rule, any existing `NetworkPolicies` and `BaselineAdminNetworkPolicy` (BANP) in the cluster will be intentionally skipped from evaluation. If the match is an ANP `pass` rule, then evaluation moves from tier 1 of the ACLs to tier 2 where `NetworkPolicy` is evaluated.

Author: I don't know what you mean. I mention it above.

> If traffic matches an ANP rule, the rules in that ANP will be evaluated first. If the match is an ANP `allow` or `deny` rule, any existing `NetworkPolicies` and `BaselineAdminNetworkPolicy` (BANP) in the cluster will be intentionally skipped from evaluation. If the match is an ANP `pass` rule, then evaluation moves from tier 1 of the ACLs to tier 2 where `NetworkPolicy` is evaluated.

Reviewer: please mention the exact sentence? I cannot see the aspects of moving from tier2 to tier3, only:
```
If the match is an ANP `pass` rule, then evaluation moves from tier 1 of the ACLs to tier 2 where `NetworkPolicy` is evaluated.
```
where is the aspect about moving from NP to BANP? Sorry if I have missed that but that was my question.

include::modules/nw-ovn-k-adminnetwork-policy.adoc[leveloffset=+1]

[discrete]

> [id="additional-resources_ANP"]

Reviewer 3: Don't need an ID for a `.` style:
```suggestion
```

.Additional resources
* xref:../../nodes/clusters/nodes-cluster-enabling-features.adoc#nodes-cluster-enabling-features[Enabling features using feature gates]

> * link:https://network-policy-api.sigs.k8s.io/[sig-network-policy-api] working group

Reviewer 3: Just a suggestion:
```suggestion
* link:https://network-policy-api.sigs.k8s.io/[Network Policy API Working Group]
```
or even just:
```suggestion
* link:https://network-policy-api.sigs.k8s.io/[Network Policy API]
```

Author: I like the first one.
